io/fastq/fastq.go

```diff
@@ -0,0 +1,300 @@
+// This is NOT a hard limit. To set a hard limit on bytes read use a
+// io.LimitReader to wrap the reader passed to the Parser.
+func (parser *Parser) ParseByteLimited(byteLimit int64) (fastqs []Fastq, bytesRead int64, err error) {
+	for bytesRead < byteLimit {
+		fastq, n, err := parser.ParseNext()
+		bytesRead += n
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				err = nil // EOF not treated as parsing error.
+			}
+			return fastqs, bytesRead, err
+		}
+		fastqs = append(fastqs, fastq)
+	}
+	return fastqs, bytesRead, nil
+}
+
+// ParseNext reads next fastq genome in underlying reader and returns the result
+// and the amount of bytes read during the call.
+// ParseNext only returns an error if it:
+//   - Attempts to read and fails to find a valid fastq sequence.
+//   - Returns reader's EOF if called after reader has been exhausted.
+//   - If a EOF is encountered immediately after a sequence with no newline ending.
+//     In this case the Fastq up to that point is returned with an EOF error.
+//
+// It is worth noting the amount of bytes read are always right up to before
+// the next fastq starts which means this function can effectively be used
+// to index where fastqs start in a file or string.
+//
+// ParseNext is simplified for fastq files from fasta files. Unlike fasta
+// files, fastq always have 4 lines following each other - not variable with
+// a line limit of 80 like fasta files have. So instead of a for loop, you
+// can just parse 4 lines at once.
+func (parser *Parser) ParseNext() (Fastq, int64, error) {
+	if _, err := parser.reader.Peek(1); err != nil {
+		// Early return on error. Probably will be EOF.
+		return Fastq{}, 0, err
+	}
+
+	// More general case of error handling.
+	handleErr := func(err error) error {
+		isEOF := errors.Is(err, io.EOF)
+		if errors.Is(err, bufio.ErrBufferFull) {
+			// Buffer size too small to read fastq line.
+			return fmt.Errorf("line %d too large for buffer, use larger maxLineSize: %w", parser.line+1, err)
+		} else if !isEOF {
+			return err // Unexpected error.
+		}
+		return nil
+	}
+
+	// Initialization of parser state variables.
+	var (
+		// Parser looks for a line starting with '@'
+		// that contains the next fastq sequence identifier.
+		lookingForIdentifier   = true
+		seqIdentifier, quality string
+		optionals              map[string]string
+		sequence, line         []byte
+		err                    error
+		totalRead              int64
+	)
+
+	// parse identifier
+	line, err = parser.reader.ReadSlice('\n')
+	totalRead += int64(len(line))
+	parser.line++
+	if handleErr(err) != nil {
+		return Fastq{}, totalRead, handleErr(err)
+	}
+	line = line[:len(line)-1] // Exclude newline delimiter.
+	if string(line)[0] == '@' {
+		lookingForIdentifier = false
+	}
+	seqIdentifier = strings.Split(string(line), " ")[0][1:]
+	optionals = make(map[string]string)
+	for _, optionalDatum := range strings.Split(string(line), " ")[1:] {
```

Comment:
We already call `strings.Split(string(line), " ")` in line 177 so I believe an initial optimization could be to do something like this and reuse the output from `strings.Split`:
```go
	splits := strings.Split(string(line), " ")
	seqIdentifier = splits[0][1:]
	optionals = make(map[string]string)
	for _, optionalDatum := range splits[1:] {
		optionalKey := strings.Split(optionalDatum, "=")[0]
		optionalValue := strings.Split(optionalDatum, "=")[1]
		optionals[optionalKey] = optionalValue
	}
```

Some benchmarks here -also results in less memory usage- 👍 
```
% go test -bench . -benchmem
goos: darwin
goarch: arm64
pkg: github.com/TimothyStiles/poly/io/fastq
BenchmarkParse-8   	  138547	      8812 ns/op	   74432 B/op	      93 allocs/op
PASS
ok  	github.com/TimothyStiles/poly/io/fastq	2.233s
% go test -bench . -benchmem
goos: darwin
goarch: arm64
pkg: github.com/TimothyStiles/poly/io/fastq
BenchmarkParse-8   	  145696	      7622 ns/op	   73088 B/op	      85 allocs/op
PASS
ok  	github.com/TimothyStiles/poly/io/fastq	2.457s
```

We can achieve the same in the `optionalDatum` block:
```go
	line = line[:len(line)-1] // Exclude newline delimiter.
	if string(line)[0] == '@' {
		lookingForIdentifier = false
	}
	lineSplits := strings.Split(string(line), " ")
	seqIdentifier = lineSplits[0][1:]
	optionals = make(map[string]string)
	for _, optionalDatum := range lineSplits[1:] {
		optionalSplits := strings.Split(optionalDatum, "=")
		optionalKey := optionalSplits[0]
		optionalValue := optionalSplits[1]
		optionals[optionalKey] = optionalValue
	}
```

And results in the following (reduces mem allocs by almost 50%, speed around 30%):
```
% go test -bench . -benchmem
goos: darwin
goarch: arm64
pkg: github.com/TimothyStiles/poly/io/fastq
BenchmarkParse-8   	  173823	      6780 ns/op	   72192 B/op	      57 allocs/op
PASS
ok  	github.com/TimothyStiles/poly/io/fastq	2.774s
```